First version docker compose one poller one container

rahulguptajss opened the pull request.

cgrinds commented on cmd/tools/generate/generate.go:
```diff
@@ -20,11 +29,55 @@ var systemdCmd = &cobra.Command{
+	color.DetectConsole("")
+	println("Docker Compose file generated at harvest home " + color.Colorize("docker-compose.yml", color.Green))
+	// Create the file
+	f, err := os.Create("docker-compose.yml")
```
probably better to dump to stdout instead of overwriting file. simple enough for the operator to redirect to docker-compse.yml. You could mention that in your println

rahulguptajss replied:
done.

cgrinds commented on cmd/tools/generate/generate.go:
```diff
@@ -20,11 +29,55 @@ var systemdCmd = &cobra.Command{
+		return
+	}
+	conf.IsDocker = true
+	for k, _ := range *conf.Config.Pollers {
```
nicer to write these in the same order as listed in the yaml - straightforward with the `PollersOrdered` member of HarvestConfig

rahulguptajss replied:
done.

cgrinds commented on docker/onePollerPerContainer/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+VOLUME     [ "/var/log/harvest" ]
+
+ENTRYPOINT ["/usr/local/bin/dumb-init", "--", "./docker-entrypoint.sh"]
```
per above, should call poller directly passing args

rahulguptajss replied:
done.

rahulguptajss commented:
> Thanks @rahulguptajss Nice start. Can you add a README.md with instructions on how to build/use?
> 
> When I try to build the docker image it failed with below, presumably a small environmental difference
> 
> ```
> [+] Building 1.6s (15/19)                                                                                                                                      
>  => [internal] load build definition from Dockerfile                                                                                                      0.0s
>  => => transferring dockerfile: 947B                                                                                                                      0.0s
>  => [internal] load .dockerignore                                                                                                                         0.0s
>  => => transferring context: 2B                                                                                                                           0.0s
>  => [internal] load metadata for docker.io/library/alpine:latest                                                                                          1.1s
>  => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                     1.1s
>  => CACHED [stage-1 1/4] FROM docker.io/library/alpine:latest@sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0                     0.0s
>  => [internal] load build context                                                                                                                         0.0s
>  => => transferring context: 1.02kB                                                                                                                       0.0s
>  => [golangalpine  1/10] FROM docker.io/library/golang:1.16-alpine@sha256:34e3951701d7cc4153ca322933ed82edf8575af0d3f5c40362dca3b3c2bc425e                0.0s
>  => CANCELED [stage-1 2/4] RUN apk add bash                                                                                                               0.4s
>  => CACHED [golangalpine  2/10] RUN apk add curl && apk add --virtual build-dependencies build-base gcc                                                   0.0s
>  => CACHED [golangalpine  3/10] RUN echo 2.0                                                                                                              0.0s
>  => CACHED [golangalpine  4/10] RUN echo nightly                                                                                                          0.0s
>  => CACHED [golangalpine  5/10] WORKDIR /opt/home                                                                                                         0.0s
>  => CACHED [golangalpine  6/10] RUN mkdir -p /opt/harvest                                                                                                 0.0s
>  => [golangalpine  7/10] COPY . .                                                                                                                         0.0s
>  => ERROR [golangalpine  8/10] RUN make build VERSION=2.0 RELEASE=nightly
> ```

Hmm, could be cache issue. Could you try with below commands from harvest root.


docker build -f docker/onePollerPerContainer/Dockerfile -t harvest:latest  . --no-cache

make build
bin/harvest generate docker > docker-compose.yml

docker-compose -f docker-compose.yml up -d

rahulguptajss commented on docker/onePollerPerContainer/docker-compose.tmpl:
```diff
@@ -0,0 +1,19 @@
+    {{end}}
+    stdin_open: true # docker run -i
+    tty: true        # docker run -t
+    command: ' --poller {{ .PollerName }} --loglevel 2 {{if .Port }} --promPort {{ .Port }} {{ end }} --config ./harvest.yml'
```
Good point. I added the file mount in docker-compose.

rahulguptajss commented on docker/onePollerPerContainer/docker-entrypoint.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+bin/poller "$@"
+`exec /bin/sleep infinity`
```
done.

rahulguptajss commented:
Is it a good idea to generate k8 artificats from docker-compose itself using Kompose?

cgrinds commented:
```
~/o/harvest (rg2-docker-compose|…) $ docker build -f docker/onePollerPerContainer/Dockerfile -t harvest:latest  . --no-cache
[+] Building 27.1s (17/19)                                                                                                                                     
 => [internal] load build definition from Dockerfile                                                                                                      0.0s
 => => transferring dockerfile: 37B                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                         0.0s
 => => transferring context: 34B                                                                                                                          0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                          0.3s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                     0.3s
 => [internal] load build context                                                                                                                         0.3s
 => => transferring context: 2.99MB                                                                                                                       0.3s
 => CACHED [golangalpine  1/10] FROM docker.io/library/golang:1.16-alpine@sha256:34e3951701d7cc4153ca322933ed82edf8575af0d3f5c40362dca3b3c2bc425e         0.0s
 => CACHED [stage-1 1/4] FROM docker.io/library/alpine:latest@sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0                     0.0s
 => [golangalpine  2/10] RUN apk add curl && apk add --virtual build-dependencies build-base gcc                                                         10.2s
 => [stage-1 2/4] RUN apk add bash                                                                                                                        2.1s
 => [golangalpine  3/10] RUN echo 2.0                                                                                                                     0.3s 
 => [golangalpine  4/10] RUN echo nightly                                                                                                                 0.3s 
 => [golangalpine  5/10] WORKDIR /opt/home                                                                                                                0.0s 
 => [golangalpine  6/10] RUN mkdir -p /opt/harvest                                                                                                        0.3s 
 => [golangalpine  7/10] COPY . .                                                                                                                         0.3s 
 => [golangalpine  8/10] RUN make build VERSION=2.0 RELEASE=nightly                                                                                      14.7s 
 => [golangalpine  9/10] RUN cp -a /opt/home/harvest.yml /opt/harvest/harvest.yml                                                                         0.3s 
 => ERROR [golangalpine 10/10] RUN cp -aR bin /opt/home/conf /opt/home/docs /opt/home/grafana /opt/home/docker/onePollerPerContainer/docker-entrypoint.s  0.4s 
------                                                                                                                                                         
 > [golangalpine 10/10] RUN cp -aR bin /opt/home/conf /opt/home/docs /opt/home/grafana /opt/home/docker/onePollerPerContainer/docker-entrypoint.sh /opt/harvest &&    chmod +x /opt/harvest/docker-entrypoint.sh:                                                                                                             
#17 0.310 cp: can't stat '/opt/home/docker/onePollerPerContainer/docker-entrypoint.sh': No such file or directory                                              
------
executor failed running [/bin/sh -c cp -aR bin $BUILD_DIR/conf $BUILD_DIR/docs $BUILD_DIR/grafana $BUILD_DIR/docker/onePollerPerContainer/docker-entrypoint.sh $INSTALL_DIR &&    chmod +x $INSTALL_DIR/docker-entrypoint.sh]: exit code: 1
```

It's true that file doesn't exist. I removed the cp and chmod from the `Dockerfile` and then the build succeeds.

`bin/harvest generate --config h2.yml docker > docker-compose.yml` correctly creates the `docker-compose.yml` file, but then `docker-compose -f docker-compose.yml up -d` fails with 

```
ERROR: for harvest_cluster-01  Cannot create container for service cluster-01: invalid mount config for type "bind": invalid mount path: 'h2.yml' mount path must be absolute
```

The instructions for how to start can be changed to `docker compose up -d`

rahulguptajss commented:
> ```
> ~/o/harvest (rg2-docker-compose|…) $ docker build -f docker/onePollerPerContainer/Dockerfile -t harvest:latest  . --no-cache
> [+] Building 27.1s (17/19)                                                                                                                                     
>  => [internal] load build definition from Dockerfile                                                                                                      0.0s
>  => => transferring dockerfile: 37B                                                                                                                       0.0s
>  => [internal] load .dockerignore                                                                                                                         0.0s
>  => => transferring context: 34B                                                                                                                          0.0s
>  => [internal] load metadata for docker.io/library/alpine:latest                                                                                          0.3s
>  => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                     0.3s
>  => [internal] load build context                                                                                                                         0.3s
>  => => transferring context: 2.99MB                                                                                                                       0.3s
>  => CACHED [golangalpine  1/10] FROM docker.io/library/golang:1.16-alpine@sha256:34e3951701d7cc4153ca322933ed82edf8575af0d3f5c40362dca3b3c2bc425e         0.0s
>  => CACHED [stage-1 1/4] FROM docker.io/library/alpine:latest@sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0                     0.0s
>  => [golangalpine  2/10] RUN apk add curl && apk add --virtual build-dependencies build-base gcc                                                         10.2s
>  => [stage-1 2/4] RUN apk add bash                                                                                                                        2.1s
>  => [golangalpine  3/10] RUN echo 2.0                                                                                                                     0.3s 
>  => [golangalpine  4/10] RUN echo nightly                                                                                                                 0.3s 
>  => [golangalpine  5/10] WORKDIR /opt/home                                                                                                                0.0s 
>  => [golangalpine  6/10] RUN mkdir -p /opt/harvest                                                                                                        0.3s 
>  => [golangalpine  7/10] COPY . .                                                                                                                         0.3s 
>  => [golangalpine  8/10] RUN make build VERSION=2.0 RELEASE=nightly                                                                                      14.7s 
>  => [golangalpine  9/10] RUN cp -a /opt/home/harvest.yml /opt/harvest/harvest.yml                                                                         0.3s 
>  => ERROR [golangalpine 10/10] RUN cp -aR bin /opt/home/conf /opt/home/docs /opt/home/grafana /opt/home/docker/onePollerPerContainer/docker-entrypoint.s  0.4s 
> ------                                                                                                                                                         
>  > [golangalpine 10/10] RUN cp -aR bin /opt/home/conf /opt/home/docs /opt/home/grafana /opt/home/docker/onePollerPerContainer/docker-entrypoint.sh /opt/harvest &&    chmod +x /opt/harvest/docker-entrypoint.sh:                                                                                                             
> #17 0.310 cp: can't stat '/opt/home/docker/onePollerPerContainer/docker-entrypoint.sh': No such file or directory                                              
> ------
> executor failed running [/bin/sh -c cp -aR bin $BUILD_DIR/conf $BUILD_DIR/docs $BUILD_DIR/grafana $BUILD_DIR/docker/onePollerPerContainer/docker-entrypoint.sh $INSTALL_DIR &&    chmod +x $INSTALL_DIR/docker-entrypoint.sh]: exit code: 1
> ```
> 
> It's true that file doesn't exist. I removed the cp and chmod from the `Dockerfile` and then the build succeeds.
> 
> `bin/harvest generate --config h2.yml docker > docker-compose.yml` correctly creates the `docker-compose.yml` file, but then `docker-compose -f docker-compose.yml up -d` fails with
> 
> ```
> ERROR: for harvest_cluster-01  Cannot create container for service cluster-01: invalid mount config for type "bind": invalid mount path: 'h2.yml' mount path must be absolute
> ```
> 
> The instructions for how to start can be changed to `docker compose up -d`

Ah right. Removed docker entry and didnt test creating new docker file.

rahulguptajss commented:
> bin/harvest generate --config h2.yml docker > docker-compose.yml

Updated dockerfile. As per documentation, Can you try command like below. Volume bind needs absolute path.

bin/harvest generate --config ./h2.yml docker > docker-compose.yml

rahulguptajss commented:
Made readme more verbose.

docker-compose -f docker-compose.yml up -d --remove-orphans

cgrinds commented:
> bin/harvest generate --config ./h2.yml docker > docker-compose.yml

That works but since we're generating the file, we should make sure we write the expected form.
Instead of `source: h2.yml` we should `source: ./h2.yml` by stating the file in the current directory

rahulguptajss commented:
> > bin/harvest generate --config ./h2.yml docker > docker-compose.yml
> 
> That works but since we're generating the file, we should make sure we write the expected form.
> Instead of `source: h2.yml` we should `source: ./h2.yml` by stating the file in the current directory

How would writing filepath.Abs(configpath) sounds?

rahulguptajss commented:
> > > bin/harvest generate --config ./h2.yml docker > docker-compose.yml
> > 
> > 
> > That works but since we're generating the file, we should make sure we write the expected form.
> > Instead of `source: h2.yml` we should `source: ./h2.yml` by stating the file in the current directory
> 
> How would writing filepath.Abs(configpath) sounds?

Added absolute path for vol mapping.

cgrinds commented:
Thanks, that should work. I'm hitting https://github.com/docker/compose-cli/issues/1811 but the [workaround](https://github.com/docker/compose-cli/issues/1811#issuecomment-863052106) makes it work.
